pkg/querier/frontend/frontend_test.go

```diff
@@ -251,8 +255,72 @@ func TestFrontendCheckReady(t *testing.T) {
 	}
 }
 
-func testFrontend(t *testing.T, config Config, handler http.Handler, test func(addr string), matchMaxConcurrency bool) {
+func TestFrontend_LogsSlowQueriesFormValues_Issue3111(t *testing.T) {
+	// Create an HTTP server listening locally. This server mocks the downstream
+	// Prometheus API-compatible server.
+	downstreamListen, err := net.Listen("tcp", "localhost:0")
+	require.NoError(t, err)
+
+	downstreamServer := http.Server{
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			_, err := w.Write([]byte(responseBody))
+			require.NoError(t, err)
+		}),
+	}
+
+	defer downstreamServer.Shutdown(context.Background()) //nolint:errcheck
+	go downstreamServer.Serve(downstreamListen)           //nolint:errcheck
+
+	// Configure the query-frontend with the mocked downstream server.
+	config := defaultFrontendConfig()
+	config.LogQueriesLongerThan = 1 * time.Microsecond
+	config.DownstreamURL = fmt.Sprintf("http://%s", downstreamListen.Addr())
+
+	var buf bytes.Buffer
+	l := log.NewLogfmtLogger(log.NewSyncWriter(&buf))
+
+	test := func(addr string) {
+		data := url.Values{}
+		data.Set("test", "form")
+		data.Set("issue", "3111")
+
+		req, _ := http.NewRequest(http.MethodPost, fmt.Sprintf("http://%s/", addr), strings.NewReader(data.Encode()))
```

Comment:
It won't work for `GET` method as https://pkg.go.dev/net/http#Request.ParseForm says:
```
// For POST, PUT, and PATCH requests, it also reads the request body, parses it
// as a form and puts the results into both r.PostForm and r.Form. Request body
// parameters take precedence over URL query string values in r.Form.
```

I've added assertion for parameters in request URL and could add similar test for `GET` method but I don't know how much of value it will have.